intel-sgx/async-usercalls/src/batch_drop.rs

```diff
@@ -0,0 +1,126 @@
+use crate::provider_core::ProviderCore;
+use ipc_queue::Identified;
+use std::cell::RefCell;
+use std::mem;
+use std::os::fortanix_sgx::usercalls::alloc::{User, UserSafe};
+use std::os::fortanix_sgx::usercalls::raw::{Usercall, UsercallNrs};
+
+pub trait BatchDroppable: private::BatchDroppable {}
+impl<T: private::BatchDroppable> BatchDroppable for T {}
+
+/// Drop the given value at some point in the future (no rush!). This is useful
+/// for freeing userspace memory when we don't particularly care about when the
+/// buffer is freed. Multiple `free` usercalls are batched together and sent to
+/// userspace asynchronously. It is also guaranteed that the memory is freed if
+/// the current thread exits before there is a large enough batch.
+///
+/// This is mainly an optimization to avoid exitting the enclave for each
+/// usercall. Note that even when sending usercalls asynchronously, if the
+/// usercall queue is empty we still need to exit the enclave to signal the
+/// userspace that the queue is not empty anymore. The batch send would send
+/// multiple usercalls and notify the userspace at most once.
+pub fn batch_drop<T: BatchDroppable>(t: T) {
+    t.batch_drop();
+}
+
+mod private {
+    use super::*;
+
+    const BATCH_SIZE: usize = 8;
+
+    struct BatchDropProvider {
+        core: ProviderCore,
+        deferred: Vec<Identified<Usercall>>,
+    }
+
+    impl BatchDropProvider {
+        pub fn new() -> Self {
+            Self {
+                core: ProviderCore::new(None),
+                deferred: Vec::with_capacity(BATCH_SIZE),
+            }
+        }
+
+        fn make_progress(&self, deferred: &[Identified<Usercall>]) -> usize {
+            let sent = self.core.try_send_multiple_usercalls(deferred);
+            if sent == 0 {
+                self.core.send_usercall(deferred[0]);
```

Comment:
I think there's a potential runtime error here if someone calls `make_progress` with an empty slice for deferred. I don't think there's currently a place that calls this function with an empty slice, but someone could change the code in the future.

Reply:
Good catch!